python/mxnet/gluon/trainer.py

```diff
@@ -170,11 +178,61 @@ def step(self, batch_size, ignore_stale_grad=False):
         if not self._kv_initialized:
             self._init_kvstore()
 
+        self._allreduce(batch_size)
+        self._update(batch_size, ignore_stale_grad)
+
+    def allreduce(self, batch_size):
+        """For each parameter, reduce the gradients from different contexts. Should be called after
+        `autograd.grad`, outside of `record()` scope, and before `trainer.update`.
+        """
+        if not self._kv_initialized:
+            self._init_kvstore()
+        assert not self._update_on_kvstore, 'allreduce() when parameters are updated on kvstore ' \
+                                            'is not supported. Try setting `update_on_kvstore` ' \
+                                            'to False when creating trainer.'
+
+        self._allreduce(batch_size)
+
+    def _allreduce(self, batch_size):
+        self._optimizer.rescale_grad = self._scale / batch_size
+
+        if self._kvstore:
+            for i, param in enumerate(self._params):
+                if param.grad_req != 'null':
+
+                    self._kvstore.push(i, param.list_grad(), priority=-i)
+
+                    if not self._update_on_kvstore:
+                        self._kvstore.pull(i, param.list_grad(), priority=-i)
+
+    def update(self, batch_size, ignore_stale_grad=False):
+        """Makes one step of parameter update. Should be called after
+        `autograd.grad` and outside of `record()` scope.
+
+        Parameters
+        ----------
+        batch_size : int
+            Batch size of data processed. Gradient will be normalized by `1/batch_size`.
+            Set this to 1 if you normalized loss manually with `loss = mean(loss)`.
+        ignore_stale_grad : bool, optional, default=False
+            If true, ignores Parameters with stale gradient (gradient that has not
+            been updated by `backward` after last step) and skip update.
+        """
+        if not self._kv_initialized:
```

Comment:
Why do we need to initialize again? Is it supposed to be initialized already since allreduce is called?